influxql/call_iterator.go

```diff
@@ -495,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
 	return []FloatPoint{{Time: ZeroTime, Value: float64(a[len(a)/2].Value)}}
 }
 
+// newModeIterator returns an iterator for operating on a mode() call.
+func NewModeIterator(input Iterator, opt IteratorOptions) (Iterator, error) {
+
+	switch input := input.(type) {
+	case FloatIterator:
+		createFn := func() (FloatPointAggregator, FloatPointEmitter) {
+			fn := NewFloatSliceFuncReducer(FloatModeReduceSlice)
```

Comment:
I thought of this for the histogram function https://github.com/influxdata/influxdb/issues/3674 as this will give the entire frequency spectrum. The reason for opting against it for mode was that it takes an extra map of size same as the slice itself (i prefer to avoid hash map where there is a possibility for lots of uniques keys). Also, I was hoping to improve on this with the one mentioned here [sublinear-algorithm-for-calculating-the-mode-of-an-array](https://www.quora.com/What-is-the-most-efficient-algorithm-for-calculating-the-mode-of-an-array-of-integers/answer/Anders-Kaseorg) ,which needs the entire sorted slice as well.  
I will try out the custom reducer and post the result here and you can make a call then.